src/drivers/imx/dummy-dma.c

```diff
@@ -45,54 +160,358 @@ static int dummy_dma_status(struct dma_chan_data *channel,
+	if (channel->cb_type & DMA_CB_TYPE_COPY)
+		channel->cb(channel->cb_data, DMA_CB_TYPE_COPY, &next);
+
 	return 0;
 }
 
+/**
+ * \brief Initialize the driver
+ * \param[in] The preallocated DMA controller structure
+ * \return 0 on success, a negative value on error
+ *
+ * This function must be called before any other will work. Calling functions
+ * such as dma_channel_get() without a successful dma_probe() is undefined
+ * behavior.
+ */
 static int dummy_dma_probe(struct dma *dma)
 {
+	struct dma_chan_data *dma_chans;
+	struct dma_chan_pdata *chanp;
+	int i;
+
+	if (dma_get_drvdata(dma)) {
+		trace_dummydma_error("dummy-dmac %d already created!",
+				     dma->plat_data.id);
+		return -EEXIST; /* already created */
+	}
+
+	dma_chans = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			    SOF_MEM_CAPS_RAM,
+			    dma->plat_data.channels * sizeof(dma_chans[0]));
+	if (!dma_chans) {
+		trace_dummydma_error("dummy-dmac %d: Out of memory!",
+				     dma->plat_data.id);
+		return -ENOMEM;
+	}
+
+	chanp = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			SOF_MEM_CAPS_RAM,
+			dma->plat_data.channels * sizeof(chanp[0]));
+	if (!chanp) {
+		rfree(dma_chans);
+		trace_dummydma_error("dummy-dmac %d: Out of memory!",
+				     dma->plat_data.id);
+		return -ENOMEM;
+	}
+
+	dma_set_drvdata(dma, dma_chans);
+
+	for (i = 0; i < dma->plat_data.channels; i++) {
+		dma_chans[i].dma = dma;
+		dma_chans[i].index = i;
+		dma_chans[i].status = COMP_STATE_INIT;
+		dma_chans[i].private = &chanp[i];
+	}
+
+	atomic_init(&dma->num_channels_busy, 0);
+
 	return 0;
 }
 
+/**
+ * \brief Free up all memory and resources used by this driver
+ * \param[in] dma The DMA controller structure belonging to this driver
+ *
+ * This function undoes everything that probe() did. All channels that were
+ * returned via dma_channel_get() become invalid and further usage of them is
+ * undefined behavior. dma_channel_put() is automatically called on all
+ * channels.
+ *
+ * This function is idempotent, and safe to call multiple times in a row.
+ */
 static int dummy_dma_remove(struct dma *dma)
 {
+	struct dma_chan_data *chans = dma_get_drvdata(dma);
+
+	tracev_dummydma("dummy_dma %d -> remove", dma->plat_data.id);
+	if (!chans)
+		return 0;
+
+	rfree(dma_chan_get_data(&chans[0]));
+	rfree(chans);
+	dma_set_drvdata(dma, NULL);
 	return 0;
 }
 
+/**
+ * \brief Get DMA copy data sizes
+ * \param[in] channel DMA channel on which we're interested of the sizes
+ * \param[out] avail How much data the channel can deliver if copy() is called
+ *		     now
+ * \param[out] free How much data can be copied to the host via this channel
+ *		    without going over the buffer size
+ * \return 0 on success, -EINVAL if a configuration error is detected
+ */
 static int dummy_dma_get_data_size(struct dma_chan_data *channel,
 				   uint32_t *avail, uint32_t *free)
 {
+	/* TODO remove hardcode, if possible */
+	switch (channel->direction) {
+	case DMA_DIR_HMEM_TO_LMEM:
+		*avail = 384;
```

Comment:
Yes, this is ugly. We need to explain what exactly is 384. Maybe write it as a result of a multiplication.

Reply:
Ack, this should not be hard coded - it should come from topology or some from driver hw_params() ?

Reply:
Will see about it. I think I could actually just add a computation for the size for 1 milisecond. OR just get the actual buffer size (the buffers between components is also 384 bytes, I technically cannot even copy more than that anyway) which is just defined somewhere else. Which sounds better? @lgirdwood

Reply:
Actually for accessing the buffer sizes in the actual pipeline I would do an inversion of dependencies or something. Should I instead calculate the available data based on the next DMA SG elem? (if you agree I will test if this doesn't break things)

Reply:
@paulstelian97 all buffer sizes and processing sizes are all passed in via the topology pipeline config. You should be able to read this data straight ouf of topology instead of hard coding it. Btw, if 384 is your only supported and validated use case, then you could safely reject any topology config that falls outside 384 until it is supported.

Reply:
@lgirdwood Thanks, where should I look exactly? Is there a possibility to determine this size without modifying host.c to pass it to the driver?

Or, would it be a sane option to just pass the size in the next elem within the dma_sg_config?

Reply:
@paulstelian97 Buffer size is passed through dma_set_config in dma_sg_elem_array (count * elem->size). You can save it in your DMA private data.